### PR TITLE
[RHCLOUD-34513] Migrate cross account requests 

### PR DESCRIPTION
**Pull request by lpichler**
## Link(s) to Jira
- https://issues.redhat.com/browse/RHCLOUD-34513

## Description of Intent of Change(s)
- Migrate cross account requests which are in approved state.

## Local Testing
How can the feature be exercised?
How can the bug be exploited and fix confirmed?
Is any special local setup required?

## Checklist
- [ ] if API spec changes are required, is the spec updated?
- [ ] are there any pre/post merge actions required? if so, document here.
- [ ] are theses changes covered by unit tests?
- [ ] if warranted, are documentation changes accounted for?
- [ ] does this require migration changes?
  - [ ] if yes, are they backwards compatible?
- [ ] is there known, direct impact to dependent teams/components?
  - [ ] if yes, how will this be handled?

## Secure Coding Practices Checklist Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Practices Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

### CODE REVIEW

File: rbac/management/relation_replicator/relation_replicator.py [hunk 1]
```diff
@@ -55,6 +55,7 @@ class ReplicationEventType(str, Enum):
     MIGRATE_SYSTEM_ROLE_ASSIGMENT = "migrate_system_role_assignment"
     APPROVE_CROSS_ACCOUNT_REQUEST = "approve_cross_account_request"
     EXPIRE_CROSS_ACCOUNT_REQUEST = "expire_cross_account_request"
+    MIGRATE_ACCOUNT_REQUEST = "migrate_cross_account_request"
```
Comment (coderbydesign) [inline rbac/management/relation_replicator/relation_replicator.py:58]: Minor nit for consistency:

```suggestion
    MIGRATE_CROSS_ACCOUNT_REQUEST = "migrate_cross_account_request"
```

File: rbac/migration_tool/migrate.py [hunk 2]
```diff
@@ -158,6 +160,27 @@ def migrate_data_for_tenant(tenant: Tenant, exclude_apps: list, replicator: Rela
...
+    for cross_principal in cross_principals:
+        cross_account_requests = cross_account_requests_dict.get(cross_principal.user_id)
+        for cross_account_request in cross_account_requests:
+            cross_account_roles = cross_account_request.roles.all()
+            if any(True for _ in cross_account_roles):
+                dual_write_handler = RelationApiDualWriteCrossAccessHandler(
+                    cross_principal, ReplicationEventType.MIGRATE_ACCOUNT_REQUEST
+                )
+                dual_write_handler.generate_relations_to_add_roles(cross_account_request.roles.all())
+                dual_write_handler.replicate()
```
Comment (alechenninger) [inline rbac/migration_tool/migrate.py:183]: Move atomic block to inner loop

File: rbac/migration_tool/migrate.py [hunk 3]
```diff
@@ -158,6 +160,27 @@ def migrate_data_for_tenant(tenant: Tenant, exclude_apps: list, replicator: Rela
...
+
+# The migrator does not generally deal with concurrency control,
+# but we require an atomic block due to use of select_for_update in the dual write handler.
+@transaction.atomic
+def migrate_cross_account_requests(tenant: Tenant, replicator: RelationReplicator):
+    cross_principals = Principal.objects.filter(cross_account=True, tenant=tenant)
+    cross_account_requests_dict = {
+        cross_principal.user_id: CrossAccountRequest.objects.filter(status="approved", user_id=cross_principal.user_id)
+        for cross_principal in cross_principals
+    }
```
Comment (alechenninger) [inline rbac/migration_tool/migrate.py:173]: Query for CAR/pass CAR to handler

**Review by alechenninger — APPROVED** [review]
LGTM

**Pull request merged**